azurerm/data_source_monitor_diagnostic_categories.go

```diff
@@ -0,0 +1,89 @@
+package azurerm
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/preview/monitor/mgmt/2018-03-01/insights"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+)
+
+func dataSourceArmMonitorDiagnosticCategories() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceArmMonitorDiagnosticCategoriesRead,
+		Schema: map[string]*schema.Schema{
+			"resource_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: azure.ValidateResourceID,
+			},
+
+			"logs": {
+				Type:     schema.TypeSet,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+				Computed: true,
+			},
+
+			"metrics": {
+				Type:     schema.TypeSet,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceArmMonitorDiagnosticCategoriesRead(d *schema.ResourceData, meta interface{}) error {
+	categoriesClient := meta.(*ArmClient).monitorDiagnosticSettingsCategoryClient
+	ctx := meta.(*ArmClient).StopContext
+
+	actualResourceId := d.Get("resource_id").(string)
+	// trim off the leading `/` since the CheckExistenceByID / List methods don't expect it
+	resourceId := strings.TrimPrefix(actualResourceId, "/")
+
+	// then retrieve the possible Diagnostics Categories for this Resource
+	categories, err := categoriesClient.List(ctx, resourceId)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Diagnostics Categories for Resource %q: %+v", actualResourceId, err)
+	}
+
+	if categories.Value == nil {
+		return fmt.Errorf("Error retrieving Diagnostics Categories for Resource %q: `categories.Value` was nil", actualResourceId)
+	}
+
+	d.SetId(actualResourceId)
+	val := *categories.Value
+
+	metrics := make([]string, 0)
+	logs := make([]string, 0)
+
+	for _, v := range val {
+		if v.Name == nil {
+			continue
+		}
+
+		if category := v.DiagnosticSettingsCategory; category != nil {
+			switch category.CategoryType {
+			case insights.Logs:
+				logs = append(logs, *v.Name)
+			case insights.Metrics:
+				metrics = append(metrics, *v.Name)
```

Comment:
Should we be nil checking here?

Reply:
Is the check in line 64 sufficient?

Reply:
this is caught above - so it should be fine 👍